### PR TITLE
Consensus indexer converts data to CID

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Consensus indexer converts data to CID

This PR aims to minimize data utilization in consensus tables (especially extrinsics and events).
When we upload data on the network, using auto-drive or anything that follows auto-dag-data models, this data gets indexed by the consensus indexer as the argument of the extrinsic (`args` column in `consensus.extrinsics` table).
This data can be indexed in a second table (`args` column in `consensus.events` table), if the extrinsics use is `system.remarkWithEvent`.

Therefore these changes convert the data into a CID if possible and replace the args by `{ "cid": "bafkr6i..." }` and add the CID in a new column.

- add auto-dag-data in consensus indexer and update other auto-sdk depencencies
- add cid column in extrinsics and events table (as index)
- add logic to try to parse the cid from the data and store it in db

___

### **PR Type**
enhancement, configuration changes

___

### **Description**
- Added CID handling to consensus indexer for extrinsics and events, minimizing data size by replacing args with CID where applicable.
- Updated database schema to include `cid` column and indexes in `consensus.events` and `consensus.extrinsics` tables.
- Introduced utility functions for CID parsing and hex conversion.
- Updated package dependencies to include `auto-dag-data` and updated versions of existing dependencies.
- Extended GraphQL schema to support CID field in `Extrinsic` and `Event` types.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>12 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>db.ts</strong><dd><code>Add CID parameter to database functions for extrinsics and events</code></dd></summary>
<hr>

indexers/mainnet/consensus/src/mappings/db.ts

<li>Added <code>cid</code> parameter to <code>createAndSaveExtrinsic</code> and <code>createAndSaveEvent</code> <br>functions.<br> <li> Updated function calls to include <code>cid</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-e94a4e233e3903b47e58104bbd6f3358fde52a89048f21d5b73b678f9f20b974">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>helper.ts</strong><dd><code>Implement CID parsing utility function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/src/mappings/helper.ts

<li>Imported CID-related functions from <code>auto-dag-data</code>.<br> <li> Added <code>parseDataToCid</code> function to parse data into CID.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-bb89264c0d5f9ce91dabc9829447614458ba1b4cd46cbefef00a2fc1285491c1">+18/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mappingHandlers.ts</strong><dd><code>Integrate CID parsing into extrinsic and event handling</code>&nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/src/mappings/mappingHandlers.ts

<li>Integrated <code>parseDataToCid</code> to handle extrinsics and events.<br> <li> Updated logic to replace args with CID if applicable.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-0e35352c284b64517989efc02c193bfc5af2ff26f7946cba711e0113e4440836">+35/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>types.ts</strong><dd><code>Define types for CID and modified arguments</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/src/mappings/types.ts

- Added new types `Cid`, `ModifiedArgs`, and `ParsedArgs`.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-61dcfa286a3d31e5207e2621cb3c2f5690b1322a6aa83bec043a389c2c21269f">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.ts</strong><dd><code>Add utility function for hex to Uint8Array conversion</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/src/mappings/utils.ts

- Added `hexToUint8Array` utility function.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-f07d40cdb5ad7e90221baa07b1876b2de5abe7285400740b6b164108008e4da4">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>db.ts</strong><dd><code>Add CID parameter to database functions for extrinsics and events</code></dd></summary>
<hr>

indexers/taurus/consensus/src/mappings/db.ts

<li>Added <code>cid</code> parameter to <code>createAndSaveExtrinsic</code> and <code>createAndSaveEvent</code> <br>functions.<br> <li> Updated function calls to include <code>cid</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-e96f8578e7ac4c5de4b1d33677bb41bd8c56a8bdce80cb2b52740d51051e417e">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>helper.ts</strong><dd><code>Implement CID parsing utility function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/src/mappings/helper.ts

<li>Imported CID-related functions from <code>auto-dag-data</code>.<br> <li> Added <code>parseDataToCid</code> function to parse data into CID.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-f0ffacaa27f2236bc750ead02caf99cd374df87d2a9102a7250742856250fb02">+18/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mappingHandlers.ts</strong><dd><code>Integrate CID parsing into extrinsic and event handling</code>&nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/src/mappings/mappingHandlers.ts

<li>Integrated <code>parseDataToCid</code> to handle extrinsics and events.<br> <li> Updated logic to replace args with CID if applicable.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-3086b8218297f23584fbf57779c7f481b6232a502cc3bfa52ae3a2e57f0ba920">+35/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>types.ts</strong><dd><code>Define types for CID and modified arguments</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/src/mappings/types.ts

- Added new types `Cid`, `ModifiedArgs`, and `ParsedArgs`.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-6b7f760b1718b4a0cb226663c679f375fc6ec0e90783a2e558882cc6a2ed4ee2">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.ts</strong><dd><code>Add utility function for hex to Uint8Array conversion</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/src/mappings/utils.ts

- Added `hexToUint8Array` utility function.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-df5dfa4f1a00d761bf7eac1034ce4030fb6f91c1c299b7484467368811b80f7b">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>schema.graphql</strong><dd><code>Extend GraphQL schema with CID field</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/schema.graphql

- Added `cid` field to `Extrinsic` and `Event` types.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-e7753b6e0a1bf90fe2615393561ec5d537ec434b9d0769ed17bb0f033dc60db0">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>schema.graphql</strong><dd><code>Extend GraphQL schema with CID field</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/schema.graphql

- Added `cid` field to `Extrinsic` and `Event` types.

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-6b01dc537e092c34cd4894dc7ee12d892203607144ccf70bae57a7d87cf8097c">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>init-db.sql</strong><dd><code>Add CID column and indexes to database schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/db/docker-entrypoint-initdb.d/init-db.sql

<li>Added <code>cid</code> column to <code>consensus.events</code> and <code>consensus.extrinsics</code> tables.<br> <li> Created indexes for <code>cid</code> in both tables.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-1f627d6006ab49e938644df52a983ebe6f7b43ce87c6d27ff4437719260d43e2">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Update package dependencies for consensus indexer</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/mainnet/consensus/package.json

<li>Updated dependencies for <code>auto-consensus</code>, <code>auto-dag-data</code>, and <br><code>auto-utils</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-c8bd61f945ad46076543d01efcffc638a83ad0f972cb7e371e750d152a339f18">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Update package dependencies for consensus indexer</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/taurus/consensus/package.json

<li>Updated dependencies for <code>auto-consensus</code>, <code>auto-dag-data</code>, and <br><code>auto-utils</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/976/files#diff-09e526e1d71f30e37a8956084b16eff70e0e01767e597d3ca935c3d4da6ffc34">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: indexers/mainnet/consensus/src/mappings/helper.ts [hunk 1]
```diff
@@ -27,3 +30,17 @@ export const getBlockAuthor = (block: SubstrateBlock): string => {
   }
   return DEFAULT_ACCOUNT_ID;
 };
+
+export const parseDataToCid = (data: string): ParsedArgs => {
+  let cid: Cid = undefined;
+  let modifiedArgs: ModifiedArgs = undefined;
+  try {
+    const node = decodeNode(hexToUint8Array(data));
```
Comment (clostao) [inline indexers/mainnet/consensus/src/mappings/helper.ts:38]: This parse works for the IPLD nodes sent as hex string but this is now deprecated in favor of raw binary data. This [PR](https://github.com/autonomys/auto-sdk/pull/171) implemented the change.

File: indexers/taurus/consensus/src/mappings/mappingHandlers.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+global.TextEncoder = require("util").TextEncoder;
+global.TextDecoder = require("util").TextDecoder;
```
Comment (clostao) [inline indexers/taurus/consensus/src/mappings/mappingHandlers.ts:2]: Why this?

File: indexers/mainnet/consensus/src/mappings/helper.ts (hunk 1 above)
Comment (clostao) [reply indexers/mainnet/consensus/src/mappings/helper.ts:38]: Indeed data shouldn't be considered a `string` always should be able to be `Bytes` too

Comment (marc-aurele-besner) [reply indexers/mainnet/consensus/src/mappings/helper.ts:38]: Do you have an extrinsic that has the data as per the new format for me to test?

File: indexers/taurus/consensus/src/mappings/mappingHandlers.ts (hunk 2 above)
Comment (marc-aurele-besner) [reply indexers/taurus/consensus/src/mappings/mappingHandlers.ts:2]: The environment SubQuery handler is running in, does not have TextEncoder&TextDecoder available, they needed to be imported.

**Review by clostao — APPROVED** [review]

File: indexers/mainnet/consensus/src/mappings/utils.ts [hunk 3]
```diff
@@ -48,3 +48,11 @@ export const getSortId = (
   const str2 = indexInBlock.toString().padStart(totalLength, "0");
   return `${str1}-${str2}`;
 };
+
+export const hexToUint8Array = (hex: string): Uint8Array => {
```
[the remark this answers is not in the document]
Comment (marc-aurele-besner) [reply indexers/mainnet/consensus/src/mappings/utils.ts:52]: Buffer is available, and is used now to differentiate between both methods [here](https://github.com/autonomys/astral/blob/e2e51f5d7975e7e3cf8234a658f5758ede6cc0f8/indexers/mainnet/consensus/src/mappings/helper.ts#L40)
That being said, decodeNode throws an error if I provide the data as Buffer
`Invalid DAG-PB form (Data must be bytes)`
If I convert the Buffer to Bytes, then this only works with the new format, therefore we won't be able to resolve the CID of the seed data

[the remark this answers is not in the document]
Comment (clostao) [reply indexers/mainnet/consensus/src/mappings/utils.ts:52]: I understand. It's weird that the `Buffer` won't work in `decodeNode` method. Anyways using Buffer.from(data, 'hex').buffer might work too or something like this:

```typescript
new Uint8Array(Buffer.from('hola').buffer)
```

**Pull request merged**